### ISSUE TITLE
Feature request/Idea

### ISSUE THREAD

**Issue by MichelNivard** [microsoft/LMOps#53]
Hi,

You could potentially enable multiple trainer models (say a great coding model and a great conversational model) by using a simple sum of two reverse KLD's?

θ = arg min J (θ) = arg min KL[qθ||p_1] + KL[qθ||p_2]

It would potentially counter the effect of ignoring important minor modes (if and when those exist) the the probability vector. This could arguable speed up base model training if you already have various solid base models with partial world knowledge (a specific language, or programming language).

I dont know enough math to know whether it would then complicate the gradient derivation in equation 2.

**Comment by t1101675** [comment]
I think it is a good idea to introduce two teacher models. The final gradient is basically the sum of the two terms' gradients, which is not complicated to implement. But how to decide the training data/prompts is still a problem. For example, if you use the data of a specific language, the coding model may fail to give precise feedback, which potentially distracts the student model.

**Comment by MichelNivard** [comment]
If you let the training models run X (say 20-80) tokens "ahead" of the student model you can let the local loss for the teacher models decide the mixing weights:

θ = arg min J (θ) = arg min w_1 * KL[qθ||p_1] + w_2 * KL[qθ||p_2]

where w_1 and w_2 are a function of local loss. obviously this could overfit things quite a bit... Alternatives include wright functions of the "certainty" of the trainer models, weight the one with the lower top_p (number of tokens that sum tpo 90% probability) heavier? so on a Chinese language text that would almost certainly weight a teacher model trained on Chinese heavier for example..

**Comment by donglixp** [comment]
Looking forward to the results.
